File: internal/providers/claude.go
package providers

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
)

type ClaudeProvider struct {
	apiKey string
}

func NewClaudeProvider(apiKey string) *ClaudeProvider {
	return &ClaudeProvider{apiKey: apiKey}
}

func (p *ClaudeProvider) Name() string { return "claude" }

func (p *ClaudeProvider) Models() []string {
	return []string{
		"claude-opus-4-5",
		"claude-sonnet-4-5",
		"claude-haiku-4-5-20251001",
		"claude-3-5-sonnet-20241022",
		"claude-3-5-haiku-20241022",
	}
}

// claudeMessage is the wire format for the Claude messages API.
type claudeMessage struct {
	Role    string      `json:"role"`
	Content interface{} `json:"content"` // string | []claudeBlock
}

type claudeBlock struct {
	Type      string      `json:"type"`
	Text      string      `json:"text,omitempty"`
	ID        string      `json:"id,omitempty"`
	Name      string      `json:"name,omitempty"`
	Input     interface{} `json:"input,omitempty"`
	ToolUseID string      `json:"tool_use_id,omitempty"`
	Content   interface{} `json:"content,omitempty"`
	IsError   bool        `json:"is_error,omitempty"`
}

type claudeTool struct {
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	InputSchema map[string]interface{} `json:"input_schema"`
}

type claudeRequest struct {
	Model     string          `json:"model"`
	Messages  []claudeMessage `json:"messages"`
	System    string          `json:"system,omitempty"`
	MaxTokens int             `json:"max_tokens"`
	Stream    bool            `json:"stream"`
	Tools     []claudeTool    `json:"tools,omitempty"`
}

func (p *ClaudeProvider) Stream(ctx context.Context, model string, messages []Message, systemPrompt string, maxTokens int, tools []ToolDef) (<-chan StreamChunk, error) {
	if p.apiKey == "" {
		return nil, fmt.Errorf("claude: no API key set — use /key claude <your-key>")
	}

	claudeMessages := marshalClaudeMessages(messages)
	claudeTools := marshalClaudeTools(tools)

	body, err := json.Marshal(claudeRequest{
		Model:     model,
		Messages:  claudeMessages,
		System:    systemPrompt,
		MaxTokens: maxTokens,
		Stream:    true,
		Tools:     claudeTools,
	})
	if err != nil {
		return nil, err
	}

	req, err := http.NewRequestWithContext(ctx, "POST", "https://api.anthropic.com/v1/messages", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("x-api-key", p.apiKey)
	req.Header.Set("anthropic-version", "2023-06-01")
	req.Header.Set("content-type", "application/json")

	ch := make(chan StreamChunk, 64)
	go func() {
		defer close(ch)

		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			ch <- StreamChunk{Error: err}
			return
		}
		defer resp.Body.Close()

		if resp.StatusCode == 429 {
			ch <- StreamChunk{RetryAfter: retryAfterSeconds(resp, 60)}
			return
		}
		if resp.StatusCode != 200 {
			var errBody struct {
				Error struct {
					Message string `json:"message"`
				} `json:"error"`
			}
			json.NewDecoder(resp.Body).Decode(&errBody)
			ch <- StreamChunk{Error: fmt.Errorf("claude error %d: %s", resp.StatusCode, errBody.Error.Message)}
			return
		}

		// Track in-progress tool calls being accumulated from the stream.
		type pendingTool struct {
			id    string
			name  string
			input strings.Builder
		}
		var pending []*pendingTool
		var currentTool *pendingTool

		scanner := bufio.NewScanner(resp.Body)
		scanner.Buffer(make([]byte, 1024*1024), 1024*1024)
		for scanner.Scan() {
			line := scanner.Text()
			if !strings.HasPrefix(line, "data: ") {
				continue
			}
			data := strings.TrimPrefix(line, "data: ")

			var event map[string]json.RawMessage
			if err := json.Unmarshal([]byte(data), &event); err != nil {
				continue
			}

			var evType string
			json.Unmarshal(event["type"], &evType)

			switch evType {
			case "content_block_start":
				var block struct {
					Index        int `json:"index"`
					ContentBlock struct {
						Type string `json:"type"`
						ID   string `json:"id"`
						Name string `json:"name"`
					} `json:"content_block"`
				}
				if json.Unmarshal([]byte(data), &block) == nil && block.ContentBlock.Type == "tool_use" {
					currentTool = &pendingTool{
						id:   block.ContentBlock.ID,
						name: block.ContentBlock.Name,
					}
					pending = append(pending, currentTool)
				} else {
					currentTool = nil
				}

			case "content_block_delta":
				var delta struct {
					Delta struct {
						Type     string `json:"type"`
						Text     string `json:"text"`
						PartJSON string `json:"partial_json"`
					} `json:"delta"`
				}
				if json.Unmarshal([]byte(data), &delta) == nil {
					switch delta.Delta.Type {
					case "text_delta":
						if delta.Delta.Text != "" {
							ch <- StreamChunk{Content: delta.Delta.Text}
						}
					case "input_json_delta":
						if currentTool != nil {
							currentTool.input.WriteString(delta.Delta.PartJSON)
						}
					}
				}

			case "content_block_stop":
				currentTool = nil

			case "message_stop":
				// Collect completed tool calls
				if len(pending) > 0 {
					calls := make([]ToolCall, len(pending))
					for i, pt := range pending {
						inp := pt.input.String()
						if inp == "" {
							inp = "{}"
						}
						calls[i] = ToolCall{ID: pt.id, Name: pt.name, Input: inp}
					}
					ch <- StreamChunk{Done: true, ToolCalls: calls}
				} else {
					ch <- StreamChunk{Done: true}
				}
				return
			}
		}
		ch <- StreamChunk{Done: true}
	}()

	return ch, nil
}

// marshalClaudeMessages converts the generic message history to Claude wire format.
func marshalClaudeMessages(msgs []Message) []claudeMessage {
	out := make([]claudeMessage, 0, len(msgs))
	for _, m := range msgs {
		switch {
		case m.Role == "assistant" && len(m.ToolCalls) > 0:
			var content []claudeBlock
			if m.Content != "" {
				content = append(content, claudeBlock{Type: "text", Text: m.Content})
			}
			for _, tc := range m.ToolCalls {
				var input interface{}
				json.Unmarshal([]byte(tc.Input), &input)
				if input == nil {
					input = map[string]interface{}{}
				}
				content = append(content, claudeBlock{
					Type:  "tool_use",
					ID:    tc.ID,
					Name:  tc.Name,
					Input: input,
				})
			}
			out = append(out, claudeMessage{Role: "assistant", Content: content})

		case m.Role == "tool":
			// Tool results go back as user messages with tool_result blocks.
			block := claudeBlock{
				Type:      "tool_result",
				ToolUseID: m.ToolUseID,
				Content:   m.Content,
				IsError:   m.IsError,
			}
			out = append(out, claudeMessage{Role: "user", Content: []claudeBlock{block}})

		default:
			out = append(out, claudeMessage{Role: m.Role, Content: m.Content})
		}
	}
	return out
}

func marshalClaudeTools(defs []ToolDef) []claudeTool {
	if len(defs) == 0 {
		return nil
	}
	tools := make([]claudeTool, len(defs))
	for i, d := range defs {
		props := map[string]interface{}{}
		for k, p := range d.Properties {
			props[k] = map[string]string{"type": p.Type, "description": p.Description}
		}
		tools[i] = claudeTool{
			Name:        d.Name,
			Description: d.Description,
			InputSchema: map[string]interface{}{
				"type":       "object",
				"properties": props,
				"required":   d.Required,
			},
		}
	}
	return tools
}

File: src/providers/claude.rs
use anyhow::{anyhow, Result};
use async_trait::async_trait;
use futures::StreamExt;
use serde_json::Value;
use tokio::sync::mpsc;

use super::{
    Message, Provider, StreamChunk, StreamRequest, ToolCall, ToolDef,
    ratelimit::{parse_rate_limit_state, retry_after_seconds},
};

pub struct ClaudeProvider {
    api_key: String,
}

impl ClaudeProvider {
    pub fn new(api_key: String) -> Self {
        Self { api_key }
    }
}

#[async_trait]
impl Provider for ClaudeProvider {
    fn name(&self) -> &str { "claude" }

    fn models(&self) -> Vec<String> {
        vec![
            "claude-opus-4-8".into(),
            "claude-sonnet-5".into(),
            "claude-haiku-4-5-20251001".into(),
            "claude-3-5-sonnet-20241022".into(),
            "claude-3-5-haiku-20241022".into(),
        ]
    }

    async fn count_tokens(&self, req: &StreamRequest) -> Option<usize> {
        if self.api_key.is_empty() { return None; }

        let messages = marshal_claude_messages(&req.messages);
        let tools = marshal_claude_tools(&req.tools);
        let mut body = serde_json::json!({
            "model": req.model,
            "messages": messages,
        });
        if !req.system_prompt.is_empty() {
            body["system"] = serde_json::json!([{
                "type": "text",
                "text": req.system_prompt,
            }]);
        }
        if !tools.is_empty() {
            body["tools"] = serde_json::json!(tools);
        }

        let resp = super::http_client()
            .post("https://api.anthropic.com/v1/messages/count_tokens")
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .header("content-type", "application/json")
            .json(&body)
            .send()
            .await
            .ok()?;

        if !resp.status().is_success() { return None; }
        let json: serde_json::Value = resp.json().await.ok()?;
        json["input_tokens"].as_u64().map(|n| n as usize)
    }

    async fn stream(&self, req: StreamRequest) -> Result<mpsc::Receiver<StreamChunk>> {
        if self.api_key.is_empty() {
            return Err(anyhow!("claude: no API key set — use /key claude <key>"));
        }

        let body = build_claude_request(&req);
        let body_bytes = serde_json::to_vec(&body)?;

        let client = super::http_client();
        let resp = client
            .post("https://api.anthropic.com/v1/messages")
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", "2023-06-01")
            .header("anthropic-beta", "prompt-caching-2024-07-31")
            .header("content-type", "application/json")
            .body(body_bytes)
            .send()
            .await?;

        let (tx, rx) = mpsc::channel::<StreamChunk>(64);

        if resp.status().as_u16() == 429 {
            let secs = retry_after_seconds(resp.headers(), 60);
            let _ = tx.send(StreamChunk {
                content: String::new(),
                done: false,
                error: None,
                tool_calls: vec![],
                retry_after: secs,
                rate_limit: None,
            }).await;
            return Ok(rx);
        }

        if !resp.status().is_success() {
            let status = resp.status().as_u16();
            let text = resp.text().await.unwrap_or_default();
            let msg = serde_json::from_str::<Value>(&text)
                .ok()
                .and_then(|v| v["error"]["message"].as_str().map(|s| s.to_string()))
                .unwrap_or(text);
            let _ = tx.send(StreamChunk {
                content: String::new(),
                done: false,
                error: Some(anyhow!("claude error {status}: {msg}")),
                tool_calls: vec![],
                retry_after: 0,
                rate_limit: None,
            }).await;
            return Ok(rx);
        }

        let rl = parse_rate_limit_state(resp.headers());
        let mut stream = resp.bytes_stream();

        tokio::spawn(async move {
            let mut buf = String::new();
            // Guards against a malicious/misbehaving endpoint sending an
            // unterminated line forever, which would otherwise grow `buf`
            // without bound while we wait for a '\n'.
            const MAX_LINE_BUF: usize = 4 * 1024 * 1024;

            // Pending tool calls being accumulated from the stream
            struct PendingTool {
                id: String,
                name: String,
                input: String,
            }
            let mut pending: Vec<PendingTool> = Vec::new();
            let mut current_tool_idx: Option<usize> = None;

            while let Some(chunk) = stream.next().await {
                let chunk = match chunk {
                    Ok(c) => c,
                    Err(e) => {
                        let _ = tx.send(StreamChunk {
                            content: String::new(),
                            done: false,
                            error: Some(anyhow!("{e}")),
                            tool_calls: vec![],
                            retry_after: 0,
                            rate_limit: None,
                        }).await;
                        return;
                    }
                };

                buf.push_str(&String::from_utf8_lossy(&chunk));

                if buf.len() > MAX_LINE_BUF {
                    let _ = tx.send(StreamChunk {
                        content: String::new(),
                        done: false,
                        error: Some(anyhow!("claude: response line exceeded {MAX_LINE_BUF} bytes without a newline, aborting stream")),
                        tool_calls: vec![],
                        retry_after: 0,
                        rate_limit: None,
                    }).await;
                    return;
                }

                // Process complete SSE lines
                loop {
                    match buf.find('\n') {
                        None => break,
                        Some(pos) => {
                            let line = buf[..pos].trim().to_string();
                            buf = buf[pos + 1..].to_string();

                            if !line.starts_with("data: ") {
                                continue;
                            }
                            let data = &line["data: ".len()..];

                            let event: Value = match serde_json::from_str(data) {
                                Ok(v) => v,
                                Err(_) => continue,
                            };

                            match event["type"].as_str().unwrap_or("") {
                                "content_block_start" => {
                                    if event["content_block"]["type"].as_str() == Some("tool_use") {
                                        let id = event["content_block"]["id"]
                                            .as_str().unwrap_or("").to_string();
                                        let name = event["content_block"]["name"]
                                            .as_str().unwrap_or("").to_string();
                                        pending.push(PendingTool { id, name, input: String::new() });
                                        current_tool_idx = Some(pending.len() - 1);
                                    } else {
                                        current_tool_idx = None;
                                    }
                                }
                                "content_block_delta" => {
                                    let delta = &event["delta"];
                                    match delta["type"].as_str().unwrap_or("") {
                                        "text_delta" => {
                                            if let Some(text) = delta["text"].as_str() {
                                                if !text.is_empty() {
                                                    let _ = tx.send(StreamChunk {
                                                        content: text.to_string(),
                                                        done: false,
                                                        error: None,
                                                        tool_calls: vec![],
                                                        retry_after: 0,
                                                        rate_limit: None,
                                                    }).await;
                                                }
                                            }
                                        }
                                        "input_json_delta" => {
                                            if let (Some(idx), Some(part)) = (
                                                current_tool_idx,
                                                delta["partial_json"].as_str(),
                                            ) {
                                                pending[idx].input.push_str(part);
                                            }
                                        }
                                        _ => {}
                                    }
                                }
                                "content_block_stop" => {
                                    current_tool_idx = None;
                                }
                                "message_stop" => {
                                    let calls: Vec<ToolCall> = pending.drain(..).map(|pt| ToolCall {
                                        id: pt.id,
                                        name: pt.name,
                                        input: if pt.input.is_empty() {
                                            "{}".into()
                                        } else {
                                            pt.input
                                        },
                                    }).collect();
                                    let _ = tx.send(StreamChunk {
                                        content: String::new(),
                                        done: true,
                                        error: None,
                                        tool_calls: calls,
                                        retry_after: 0,
                                        rate_limit: rl,
                                    }).await;
                                    return;
                                }
                                _ => {}
                            }
                        }
                    }
                }
            }

            let calls: Vec<ToolCall> = pending.drain(..).map(|pt| ToolCall {
                id: pt.id,
                name: pt.name,
                input: if pt.input.is_empty() { "{}".into() } else { pt.input },
            }).collect();
            let _ = tx.send(StreamChunk {
                content: String::new(),
                done: true,
                error: None,
                tool_calls: calls,
                retry_after: 0,
                rate_limit: rl,
            }).await;
        });

        Ok(rx)
    }
}

fn build_claude_request(req: &StreamRequest) -> Value {
    let messages = marshal_claude_messages(&req.messages);
    let tools = marshal_claude_tools(&req.tools);

    let system: Value = if req.system_prompt.is_empty() {
        Value::Null
    } else {
        serde_json::json!([{
            "type": "text",
            "text": req.system_prompt,
            "cache_control": { "type": "ephemeral" }
        }])
    };

    let mut body = serde_json::json!({
        "model": req.model,
        "messages": messages,
        "max_tokens": req.max_tokens,
        "stream": true,
    });

    if !system.is_null() {
        body["system"] = system;
    }
    if !tools.is_empty() {
        body["tools"] = serde_json::json!(tools);
    }

    body
}

fn marshal_claude_messages(msgs: &[Message]) -> Vec<Value> {
    let mut out = Vec::new();
    for m in msgs {
        match m.role.as_str() {
            "assistant" if !m.tool_calls.is_empty() => {
                let mut content: Vec<Value> = Vec::new();
                if !m.content.is_empty() {
                    content.push(serde_json::json!({"type": "text", "text": m.content}));
                }
                for tc in &m.tool_calls {
                    let input: Value = serde_json::from_str(&tc.input)
                        .unwrap_or(serde_json::json!({}));
                    content.push(serde_json::json!({
                        "type": "tool_use",
                        "id": tc.id,
                        "name": tc.name,
                        "input": input,
                    }));
                }
                out.push(serde_json::json!({"role": "assistant", "content": content}));
            }
            "tool" => {
                out.push(serde_json::json!({
                    "role": "user",
                    "content": [{
                        "type": "tool_result",
                        "tool_use_id": m.tool_use_id,
                        "content": m.content,
                        "is_error": m.is_error,
                    }]
                }));
            }
            _ => {
                out.push(serde_json::json!({"role": m.role, "content": m.content}));
            }
        }
    }
    out
}

fn marshal_claude_tools(defs: &[ToolDef]) -> Vec<Value> {
    defs.iter().map(|d| {
        let mut props: serde_json::Map<String, Value> = serde_json::Map::new();
        for p in &d.properties {
            props.insert(p.name.clone(), serde_json::json!({
                "type": p.ty,
                "description": p.description,
            }));
        }
        serde_json::json!({
            "name": d.name,
            "description": d.description,
            "input_schema": {
                "type": "object",
                "properties": props,
                "required": d.required,
            }
        })
    }).collect()
}
